src/lib/Syntax.hs

```diff
@@ -1119,25 +1314,32 @@ substEffTail :: SubstEnv -> Maybe Name -> EffectRow
 substEffTail _ Nothing = EffectRow mempty Nothing
 substEffTail env (Just v) = case envLookup env (v:>()) of
   Nothing -> EffectRow mempty (Just v)
-  Just (Var (v':>_)) -> EffectRow mempty (Just v')
-  Just (Eff r) -> r
+  Just (Rename v')              -> EffectRow mempty (Just v')
+  Just (SubstVal (Var (v':>_))) -> EffectRow mempty (Just v')
```

Comment:
Yeah, it'd be nice to avoid it. I think we still need it because effect tails can be substituted with ordinary atoms, via beta reduction, and those atoms might just happen to be variables, which would give the `SubstVal (Var _)` case.